rocsp/rocsp.go

```diff
@@ -201,3 +201,99 @@ func (c *Client) GetMetadata(ctx context.Context, serial string) (*Metadata, err
+// `Serial` is the stringified serial number of the response. `Body` is the
+// DER bytes of the response. If this object represents an error, `Err` will
+// be non-nil and the other entries will have their zero values.
+type ScanResponsesResult struct {
+	Serial string
+	Body   []byte
+	Err    error
+}
+
+// ScanResponses scans Redis for all OCSP responses where the serial number matches the provided pattern.
+// It returns immediately and emits results and errors on `<-chan ScanResponsesResult`. It closes the
+// channel when it is done or hits an error.
+func (c *Client) ScanResponses(ctx context.Context, serialPattern string) <-chan ScanResponsesResult {
+	pattern := fmt.Sprintf("r{%s}", serialPattern)
+	results := make(chan ScanResponsesResult)
+	go func() {
+		defer close(results)
+		var cursor uint64
+		for {
+			var keys []string
+			var err error
+			keys, cursor, err = c.rdb.Scan(ctx, cursor, pattern, 10).Result()
+			if err != nil {
+				results <- ScanResponsesResult{Err: err}
+				return
+			}
+			if cursor == 0 {
+				return
+			}
+			for _, key := range keys {
+				if len(key) != 39 || key[0:2] != "r{" || key[38:39] != "}" {
+					results <- ScanResponsesResult{Err: fmt.Errorf("malformed Redis key %q", key)}
+					return
+				}
+				serial := key[2:38]
+				val, err := c.rdb.Get(ctx, key).Result()
+				if err != nil {
+					results <- ScanResponsesResult{Err: fmt.Errorf("getting metadata: %w", err)}
+					return
+				}
+				results <- ScanResponsesResult{Serial: serial, Body: []byte(val)}
+			}
+		}
+	}()
+	return results
+}
+
+// ScanResponsesResult represents a single OCSP response entry in redis.
+// `Serial` is the stringified serial number of the response. `Metadata` is the
+// parsed metadata. If this object represents an error, `Err` will
+// be non-nil and the other entries will have their zero values.
+type ScanMetadataResult struct {
+	Serial   string
+	Metadata *Metadata
+	Err      error
+}
+
+// ScanResponses scans Redis for the metadata of all OCSP responses where the serial number matches
+// the provided pattern. It returns immediately and emits results and errors on
+// `<-chan ScanResponsesResult`. It closes the channel when it is done or hits an error.
+func (c *Client) ScanMetadata(ctx context.Context, serialPattern string) <-chan ScanMetadataResult {
+	pattern := fmt.Sprintf("m{%s}", serialPattern)
+	results := make(chan ScanMetadataResult)
+	go func() {
+		defer close(results)
+		var cursor uint64
+		for {
+			var keys []string
+			var err error
+			keys, cursor, err = c.rdb.Scan(ctx, cursor, pattern, 10).Result()
+			if err != nil {
+				results <- ScanMetadataResult{Err: err}
+				return
+			}
+			if cursor == 0 {
+				return
+			}
+			for _, key := range keys {
+				if len(key) != 39 || key[0:2] != "m{" || key[38:39] != "}" {
+					results <- ScanMetadataResult{Err: fmt.Errorf("malformed Redis key %q", key)}
+					return
+				}
+				serial := key[2:38]
+				m, err := c.GetMetadata(ctx, serial)
```

Comment:
Can just use `c.Get(ctx, key)` to avoid reconstructing the key from the serial. You already do this in `ScanResponses` above.

Reply:
Good idea!

Reply:
Actually, `c.Get` doesn't work here because I wanted to get a parsed Metadata object, not a string from Redis.

Reply:
Hmm you're right. It's unfortunate that you have to parse the serial from the key here, and then reconstruct the key from the serial inside `GetMetadata`. Or maybe what's weird is that this uses the typed `GetMetadata` while `ScanResponses` above just (lazily?) uses `Get`, because both `Get` and `GetResponse` have the same return type.

I'm not sure what the best solution -- if anything -- is here, but the asymmetry is bothersome to me.